File: ruby/easy/203_remove_linked_list_elements.rb
# 203. Remove Linked List Elements
# Easy
#
# Given the head of a linked list and an integer val, remove all the nodes
# of the linked list that has Node.val == val, and return the new head.

class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

def remove_elements(head, val)
  return nil if head.nil?

  head = head.next while head && head.val == val

  return nil if head.nil?

  prev = head
  cur = head.next
  while cur
    if cur.val == val
      prev.next = cur.next
      cur.next = nil
      cur = prev.next
    else
      cur = cur.next
      prev = prev.next
    end
  end

  head
end

File: python/easy/203_remove_linked_list_elements.py
"""
203. Remove Linked List Elements
Easy

Given the head of a linked list and an integer val, remove all the nodes
of the linked list that has Node.val == val, and return the new head.
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        if head is None:
            return None

        while head is not None and head.val == val:
            head = head.next

        if head is None:
            return None

        prev = head
        curr = head.next

        while curr is not None:
            if curr.val == val:
                prev.next = curr.next
                curr = prev.next
            else:
                curr = curr.next
                prev = prev.next

        return head
